substrate/client/consensus/aura/src/import_queue.rs

```diff
@@ -160,13 +159,100 @@ where
 
 		Ok(())
 	}
+
+	async fn check_and_report_equivocation(
+		&self,
+		slot_now: Slot,
+		slot: Slot,
+		header: &B::Header,
+		author: &AuthorityId<P>,
+		origin: &BlockOrigin,
+	) -> Result<(), Error<B>> {
+		// Don't report any equivocations during initial sync as they are most likely stale.
+		if !self.check_for_equivocation || *origin == BlockOrigin::NetworkInitialSync {
+			return Ok(())
+		}
+
+		// Check if authorship of this header is an equivocation and return a proof if so.
+		let equivocation_proof =
+			match check_equivocation(&*self.client, slot_now, slot, header, author)
+				.map_err(Error::Client)?
+			{
+				Some(proof) => proof,
+				None => return Ok(()),
+			};
+
+		info!(
+			target: LOG_TARGET,
+			"Slot author is equivocating at slot {} with headers {:?} and {:?}",
+			slot,
+			equivocation_proof.first_header.hash(),
+			equivocation_proof.second_header.hash(),
+		);
+
+		// Get the best block on which we will build and send the equivocation report.
+		let best_hash = self
+			.select_chain
+			.best_chain()
+			.await
+			.map(|h| h.hash())
+			.map_err(|e| Error::Client(e.into()))?;
+
+		// Generate a key ownership proof. we start by trying to generate the
+		// key ownership proof at the parent of the equivocating header, this
+		// will make sure that proof generation is successful since it happens
+		// during the on-going session (i.e. session keys are available in the
+		// state to be able to generate the proof). This might fail if the
+		// equivocation happens on the first block of the session, in which case
+		// its parent would be on the previous session. If generation on the
+		// parent header fails we try with best block as well.
```

Comment:
how this work if equivocating authority was part of previous session as well?
Isn't the proof produced with the wrong session-id?

Reply:
Isn't this a security hole? I mean, equivocations on first session blocks are never considered if at least one authority changes

Reply:
What is the point in trying with best_hash?

Reply:
Probably equivocation proof creation/submission should be performed AFTER block import? At that point we have the authorities in the state